doc: add partition mode hld (#2063)
projectacrn/acrn-hypervisor: doc/developer-guides/hld/hv-partitionmode.rst

.. _partition-mode-hld:

Partition mode
##############

ACRN is type-1 hypervisor that supports running multiple guest operating
systems (OS). Typically, the platform BIOS/boot-loader boots ACRN, and
ACRN loads single or multiple guest OSes. Refer to :ref:`hv-startup` for
details on the start-up flow of the ACRN hypervisor.

ACRN supports two modes of operation: Sharing mode and Partition mode.
This document describes ACRN's high-level design for Partition mode
support.

.. contents::
   :depth: 2
   :local:

Introduction
************

ACRN, in partition mode, is used to provide guests with exclusive access

> to cores, memory, cache, and IO devices. There is no resource sharing

Reviewer: Can we provide exclusive access to last level cache? If not, we may need to be more specific on 'exclusive access to cache' here.

Reviewer 2: @junjiemao1 I added this as a placeholder before we add RDT support to ACRN. Since RDT support is in our roadmap, can we keep it here for now or remove it and add later when RDT implementation is completed?

Reviewer: How about adding a note saying that the cache part is partitioned only after RDT is supported?

> to cores, memory, cache, and IO devices. There is no resource sharing

Reviewer 2: @gvancuts I agree. Peripheral devices should be the right term.

> between the guests. ACRN expects static partitioning, by developers, of

Reviewer: 'No resource sharing between the guests' is a stronger claim than what we currently support. The last level cache, memory bandwidth and peripheral bus are still shared. Can we make this statement more specific to what we partition?

Reviewer 2: Sure, will add more details to it.

the location of each guest kernel in memory. ACRN creates a copy of each
guest kernel into each of the guests' memory. Current implementation of
ACRN requires developers to specify kernel parameters for the guests as
part of guest configuration. ACRN picks up kernel parameters from guest
configuration and copies them to the corresponding guest memory.

.. figure:: images/partition-image18.png
   :align: center

ACRN set-up for guests
**********************

Cores
=====

ACRN requires the developer to specify the number of guests and the
cores dedicated for each guest. Also the developer needs to specify
the physical core used as the Boot Strap Processor (BSP) for each guest. As
the processors are brought to life in the hypervisor, it checks if they are
configured as BSP for any of the guests. If a processor is BSP of any of
the guests, ACRN proceeds to build the memory mapping for the guest,
mptable, E820 entries, and zero page for the guest. As described in
`Guest info`_, ACRN creates copies of guest kernel and kernel
parameters into guest memory. :numref:`partBSPsetup` explains these
events in chronological order.

> 

Reviewer: Can we also describe the operating mode from with the BSP starts with? I remember this is different in partition mode and sharing mode.

Reviewer 2: @junjiemao1 I think that difference is captured. In sharing mode, only the platform BSP proceeds with launching SOS. In partition mode, APs also check if they are any of the guests BSP.

e820 info for all the guests.

+------------------------+
| RAM                    |
|                        |
| 0 - 0xEFFFFH           |
+------------------------+
| RESERVED (MPTABLE)     |
|                        |
| 0xF0000H - 0x100000H   |
+------------------------+
| RAM                    |
|                        |
| 0x100000H - LOWMEM     |
+------------------------+
| RESERVED               |
+------------------------+
| PCI HOLE               |
+------------------------+
| RESERVED               |
+------------------------+

Platform info - mptable
=======================

ACRN, in partition mode, uses mptable to convey platform info to each
guest.  Using this platform information, number of cores used for each
guest, and whether the guest needs devices with INTX, ACRN builds
mptable and copies it to the guest memory. In partition mode, ACRN uses
physical APIC IDs to pass to the guests.

I/O - Virtual devices
=====================

Port I/O is supported for PCI device config space 0xcfc and 0xcf8, vUART
0x3f8, vRTC 0x70 and 0x71, and vPIC ranges 0x20/21, 0xa0/a1, and
0x4d0/4d1.  MMIO is supported for vIOAPIC. ACRN exposes a virtual

> host-bridge at B.D:F 0.0:0 to each guest. Access to 256 bytes of config

Reviewer 3: ```suggestion
host-bridge at BDF (Bus Device Function) ``0:0.0`` to each guest. Access to 256 bytes of config
```

space for virtual host bridge is emulated.

I/O - Pass-thru devices
=======================

ACRN, in partition mode, supports passing thru PCI devices on the
platform. All the pass-thru devices are exposed as child devices under
the virtual host bridge. ACRN does not support either passing thru
bridges or emulating virtual bridges. Pass-thru devices should be
statically allocated to each guest using the guest configuration. ACRN

> expects the developer to provide the virtual Bus.Device:Function (BDF)

Reviewer 3: ```suggestion
expects the developer to provide the virtual Bus:Device.Function (BDF)
```

to BDF of the physical device mapping for all the pass-thru devices as
part of each guest configuration.

Run-time ACRN support for guests
********************************

ACRN, in partition mode, supports an option to pass-thru LAPIC of the
physical CPUs to the guest. ACRN expects developers to specify if the
guest needs LAPIC pass-thru using guest configuration. When guest
configures vLAPIC as x2APIC, and if the guest configuration has LAPIC
pass-thru enabled, ACRN passes the LAPIC to the guest. Guest can access
the LAPIC hardware directly without hypervisor interception. During
runtime of the guest, this option differentiates how ACRN supports
inter-processor interrupt handling and device interrupt handling. This
will be discussed in detail in the corresponding sections.

.. figure:: images/partition-image16.png
   :align: center

Guest SMP boot flow
===================

The core APIC IDs are reported to the guest using mptable info. SMP boot

> flow is similar to sharing mode. Refer to <hld> for guest SMP

Reviewer: Please replace `<hld>` with a link to the actual section in the original HLD.

Author: I agree.  @sainath14 Where in the HLD should this link?

boot flow in ACRN.

Inter-processor Interrupt (IPI) Handling
========================================

Guests w/o LAPIC pass-thru
--------------------------

For guests without LAPIC pass-thru, IPIs between guest CPUs are handled in

> the same way as sharing mode of ACRN. Refer to <hld> for more

Reviewer 3: Where should <hld> link to?

details.

Guests w/ LAPIC pass-thru
-------------------------

ACRN supports pass-thru if and only if the guest is using x2APIC mode
for the vLAPIC. In LAPIC pass-thru mode, writes to Interrupt Command
Register (ICR) x2APIC MSR is intercepted. Guest writes the IPI info
including vector, destination APIC IDs to the ICR. Upon an IPI request
from the guest, ACRN does sanity check on the destination processors
programmed into ICR. If the destination is a valid target for the guest,
ACRN sends IPI with the same vector from ICR to the physical CPUs
corresponding to the destination processor info in ICR.

.. figure:: images/partition-image14.png
   :align: center

Pass-thru device support
========================

Configuration space access
--------------------------

ACRN provides emulation for Configuration Space Address (0xcf8) IO port
and Configuration Space Data (0xcfc) IO port for guests to access PCI
devices configuration space. Base Address registers (BAR), offsets
starting from 0x10H to 0x24H, provide the information about the
resources (IO and MMIO) used by the PCI device. ACRN passes thru reads

> and writes to the physical BAR of the pass-thru devices except for BARs.

Reviewer 3: I do not understand this sentence.

Refer to IO section below for more details.

.. figure:: images/partition-image1.png
   :align: center

DMA
---

ACRN developers need to statically define the pass-thru devices for each
guest using the guest configuration. For devices to DMA to/from guest

> memory directly, ACRN parses thru the list of pass-thru devices for each

Reviewer 3: Should it read?
```suggestion
memory directly, ACRN parses the list of pass-thru devices for each
```

guest and creates context entries in the VT-d remapping hardware. EPT
page tables created for the guest are used for VT-d page tables.

I/O
---

ACRN supports I/O for pass-thru devices with two restrictions.

1) Supports only MMIO. So requires developers to expose I/O BARs as
   not present in the guest configuration.

2) Supports only 32-bit MMIO BAR type.

As guest PCI sub-system scans the PCI bus and assigns Guest Physical
Address (GPA) to the MMIO BAR, ACRN maps GPA to the address in the
physical BAR of the pass-thru device using EPT. Following timeline chart
explains how PCI devices are assigned to guest and BARs are mapped upon
guest initialization.

> .. figure:: images/partition-image13.png

Reviewer 3: Candidate for conversion to graphviz? @dbkinder

supported.

MSI support
~~~~~~~~~~~

Guest reads/writes to PCI configuration space for configuring MSI
interrupts using address. Data and control registers are pass-thru to
the physical BAR of pass-thru device. Refer to `Configuration
space access`_ for details on how PCI configuration space is emulated.

Virtual device support
======================

ACRN provides read-only vRTC support for partition mode guests. Writes
to the data port are discarded.

For port I/O to ports other than vPIC, vRTC, or vUART, reads return 0xFF and
writes are discarded.

Interrupt delivery
==================

Guests w/o LAPIC pass-thru
--------------------------

In partition mode of ACRN, interrupts stay disabled after a vmexit.  The
processor does not take interrupts when it is executing in VMX root
mode. ACRN configures the processor to take vmexit upon external
interrupt if the processor is executing in VMX non-root mode. Upon an
external interrupt, after sending EOI to the physical LAPIC, ACRN
injects the vector into the vLAPIC of vCPU currently running on the
processor. Guests using Linux as kernel, uses vectors less than 0xECh
for device interrupts.

> .. figure:: images/partition-image20.png

Reviewer 3: @dbkinder should we use graphviz instead here?

Reviewer 3: > @gvancuts Let's get this approved and merged, then I'll come back around and convert images to graphviz as appropriate.

Works for me, I'll remove all comments I made about graphviz so we concentrate and close on the other comments.
